### PR TITLE
Accelerate Unsafe CAS Intrinsics on Power and X

### PR DESCRIPTION
**Pull request by IBMJimmyk**
Adds support for the following recognized methods:
```
CompareAndExchangeObject     //JDK11
CompareAndExchangeReference  //JDK17,21
CompareAndExchangeInt        //JDK11,17,21
CompareAndExchangeLong       //JDK11,17,21
```

Similar to their CompareAndSet counterparts, the JIT acceleration does not support CAE on static fields so the createUnsafeCASCallDiamond function was updated to also work on the CAE methods.

The accelerated CAE code was built on top of the existing accelerated CAS support on both Power and X.

Edit:
X changes are dependent on this OMR PR:
https://github.com/eclipse/omr/pull/7438
Behavior of `inlineCompareAndSwapNative` and `inlineCompareAndSwapObjectNative` in `J9TreeEvaluator.cpp` must match behavior in `willBeEvaluatedAsCallByCodeGen` in `OMRCodeGenerator.cpp`

### CODE REVIEW

**Comment by zl-wang** [comment]
are we supposed to go around the "public" (not sure about this) Unsafe interface/class and support the internal Unsafe APIs? i.e., do typical developers import jdk.internal.misc.Unsafe? In addition to recognize the former class' APIs, we need to recognize the internal newer ones as well? @tajila @vijaysun-omr  what is your take?

**Comment by vijaysun-omr** [comment]
Are you suggesting that there are "public" equivalents of the apis being discussed in this PR (that we already accelerate or plan to accelerate) ? If not, then internal (non-public) or not, it could have a bearing on performance, right ?

**Comment by zl-wang** [comment]
there are two Unsafe classes in JDK ... for example jdk21:
`src/jdk.unsupported/share/classes/sun/misc/Unsafe.java`
`src/java.base/share/classes/jdk/internal/misc/Unsafe.java`

APIs in the former class basically use the latter's for implementation. Something like below:
```
    public final boolean compareAndSwapObject(Object o, long offset,
                                              Object expected,
                                              Object x) {
        return theInternalUnsafe.compareAndSetReference(o, offset, expected, x);
    }

```

if programmers are expected (if they ever do) to use the former only, it might not be necessary to go behind it and deal with the new names.

**Comment by IBMJimmyk** [comment]
Only these ones exist in `sun/misc/Unsafe`:
```
compareAndSwapObject
compareAndSwapInt
compareAndSwapLong
```
These ones require going through `jdk/internal/misc/Unsafe.java`:
```
CompareAndExchangeObject
CompareAndExchangeReference
CompareAndExchangeInt
CompareAndExchangeLong
```

**Comment by tajila** [comment]
> we need to recognize the internal newer ones as well? @tajila @vijaysun-omr what is your take?

Most new applications and the JDK itself have switched to using jdk.internal.misc.Unsafe. sun.misc.Unsafe is still available but will likely go at some point.

One thing to remember is that in JDK8 there is only sun.misc.Unsafe.

**Comment by zl-wang** [comment]
> Most new applications and the JDK itself have switched to using jdk.internal.misc.Unsafe.

Thanks for the clarification.

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp
Comment (zl-wang) [inline runtime/compiler/p/codegen/J9TreeEvaluator.cpp:8186]: this "if" block is expected to be common to all cases (even for future byte/short CAS support). why not move/common it to after-if?

**Review by zl-wang — CHANGES_REQUESTED** [review]

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp
Comment (IBMJimmyk) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp:8186]: I can move it outside the switch block.

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp
Comment (IBMJimmyk) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp:12215]: The original problem traces back to this commit: https://github.com/eclipse-openj9/openj9/commit/111f0b0bcc55e3ae7d88900d0dbcca460e7ed875

Going from Java 8 to 11, `compareAndSwap[Object|Int|Long]` were changed to being wrappers to call Native versions with the same name. So there is both a java and native that matches the same recognized method.

Going from 11 to 17, `compareAndSetObject` and `compareAndExchangeObject` were changed from native to Java wrapper that calls the new names `compareAndSetReference` and `compareAndExchangeReference` that do the same thing.

That commit I linked to deals with `compareAndSwap[Object|Int|Long]` and `compareAndSetObject`. Acceleration for `compareAndExchangeObject` didn't exist so that case didn't need to be handled.

Based on the comment in the commit, the change explicitly went in to deal with incorrect behavior. But, it doesn't say what it is and I'm not exactly sure what the problem is either.

I copied the same check for the acceleration of `compareAndExchange[Reference|Int|Long]`. But, those are always native in every JDK level they exist in. So I can remove the check for those cases.

`compareAndSet[Reference|Int|Long]` is also always native but uses the same recognized methods as `compareAndSwap[Object|Int|Long]` so I can't remove the check for that case without updating Z and arm code which is currently future work.

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp
Comment (IBMJimmyk) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp:12220]: I can do that.

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp
Comment (IBMJimmyk) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp:12215]: I talked to @JamesKingdon and he pointed me towards this issue from around 7 years ago: https://github.com/eclipse-openj9/openj9/issues/321

From the issue the test would hang. The exact reasoning is not listed but somehow it was determined that preventing the non-native wrapped calls of the CAS methods from being turned into accelerated assembly made the problem stop. As a conservative fix, the optimization was limited to just the native methods and to leave it to the inliner to inline the wrapper methods.

File: runtime/compiler/env/j9method.cpp
Comment (hzongaro) [inline runtime/compiler/env/j9method.cpp:5828]: Was the presence of these three `Unsafe.compareAndSwap` methods in `isUnsafePut` a bug or do the callers of `isUnsafePut` expect this method to return true for those methods?

**Review by hzongaro — COMMENTED** [review]
Still reviewing, but I thought I’d give some initial feedback.

File: runtime/compiler/env/j9method.cpp
Comment (IBMJimmyk) [reply runtime/compiler/env/j9method.cpp:5548]: I can change that.

File: runtime/compiler/env/j9method.cpp
Comment (IBMJimmyk) [reply runtime/compiler/env/j9method.cpp:5828]: None of the callers of `isUnsafePut` use these three cases. They all have earlier checks for different recognized methods. The three callers are:
`UnsafeFastPath` is trying to turn Unsafe Put into a indirect store node. It doesn't handle CAS operations. An earlier check filters out the CAS recognized methods.
`truncateBooleanForUnsafeGetPut` works with boolean and none of these are booleans. An earlier check limit this to the boolean case which filters out the CAS recognized methods.
`processUnsafeAtomicCall` is only called on the recognized methods: `getAnd[Add|Set][Int|Long]`

I opted to remove the cases since it felt confusing to me. I would expect an "Unsafe Put" to be one of the various Unsafe methods with "Put" in the name that have similar properties. If anything, I would support moving `getAnd[Add|Set][Int|Long]` out of `isUnsafePut` and into its own check since those don't feel like an "Unsafe Put" to me either. But, it does currently work with how the code is written since there are checks that limit a match to the third case.

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp
Comment (IBMJimmyk) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp]: The linked issue was create due to an external test that I do not have access to. The issue also includes a simple program which could reproduce the issue but it no longer runs. The is mention in the issue of a dependency on a server that I assume is no longer running. So it does not seem to be possible to reproduce the problem.

I suspect that the original fix was only hiding an issue because it doesn't make much sense that it would have actually fixed the problem. Putting down accelerated assembly should have worked when replacing non-native Unsafe CAS call wrappers. As it would turn out, while creating this PR I encountered an issue with reference counting inside the X fast path generation for Unsafe CAS. While a different path is taken today, 7 years ago `compareAndSwapObject`, which appeared to cause the problem, went through `inlineCompareAndSwapNative` and that's where I found the problem with reference counting. It is also possible the problem was something else fixed in the last 7 years. Without being able to run the test, I can't confirm that.

**Comment by IBMJimmyk** [comment]
I added a few more fixes mostly related to problems on x that showed up after more rigorous testing.

I missed adding `jdk_internal_misc_Unsafe_compareAndExchangeObject` to one location on x which stopped the method from being converted into inlined assembly so I added that in.

Setting `TR_UseOldCompareAndSwapObject` would cause `compareAndExchange[Object|Reference]` to break on x since the older path was not updated to work with these methods. I updated them now.

Setting `TR_UseOldCompareAndSwapObject` and `TR_DisableCASInlining` would cause problems, also on x, even without my changes. This is related to hiding object compression from the codegen. I added a fix here so the compression is not hidden if `TR_DisableCASInlining` is set.

I found another location I missed adding `disableCAEIntrinsic` to so I added that in.

I got one of the native checks backwards inside `alwaysWorthInlining`. This check was for natives that already failed the inlinable JNI check so they aren't supposed to return true from `alwaysWorthInlining`.

**Comment by IBMJimmyk** [comment]
With the latest change, all the tests that I wrote to try out different edges cases are working as expected. @hzongaro, I know it has only been a few days since my last x related change but have you had a chance to continue your code review?

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp
Comment (zl-wang) [inline runtime/compiler/p/codegen/J9TreeEvaluator.cpp:8123]: i can see some contradictions here: you seemed assuming that the coming-in old/new values in registers are still compressed (if we run in compressedRef mode of course).  you are using them as if the assumption were true. however, compressed values should never exist for any substantial duration (e.g. over a GC point). usually, they are decompressed right away after loading, as you did above for `resultReg`.   that is the contradiction i referred to (i.e. old/new values passed in registers -- the nodes might have existed there and re-referenced many times for a long time -- are still compressed).

**Review by zl-wang — CHANGES_REQUESTED** [review]

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp
Comment (IBMJimmyk) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp:8123]: The reason for this inconsistency is due to this:
https://github.com/IBMJimmyk/openj9/blob/48f927706ac45e937e059a2879d2007421e4ef24/runtime/compiler/codegen/J9CodeGenerator.cpp#L655-L677

There is existing logic that causes the compressedrefs to not be decompressed before being used by codegen. I had planned to get rid of this after looking at Z since it affects that platform as well. But, I could try removing it just for Power if you think that would be better.

Comment (zl-wang) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp:8123]: i am ok with visiting it later after this merge.

**Review by zl-wang — APPROVED** [review]
LGTM now

**Comment by hzongaro** [comment]
> have you had a chance to continue your code review?

I hadn't had a chance to come back to it.  I'll take another look at it now.

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp
Comment (hzongaro) [inline runtime/compiler/p/codegen/J9TreeEvaluator.cpp:12223]: May I suggest eliminating the `break` in the middle of this case block by changing this to the following?

```
if (!disableCAEIntrinsic)
   {
   resultReg = VMinlineCompareAndSetOrExchange(node, cg, 4, true);
   return true;
   }
```

and similarly for the other new cases in this `switch`?

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp
Comment (hzongaro) [inline runtime/compiler/p/codegen/J9TreeEvaluator.cpp:8176]: Minor nit:  may I ask you to change this to a `//` style comment?

**Review by hzongaro — APPROVED** [review]
I think the changes look good.  I have a few stylistic comments, but I will approve your changes and leave it up to you whether you accept none, any or all of the suggestions.

File: runtime/compiler/optimizer/InlinerTempForJ9.cpp
Comment (IBMJimmyk) [reply runtime/compiler/optimizer/InlinerTempForJ9.cpp:413]: I'll change that.

File: runtime/compiler/optimizer/InlinerTempForJ9.cpp
Comment (IBMJimmyk) [reply runtime/compiler/optimizer/InlinerTempForJ9.cpp:433]: I'll copy the return.

File: runtime/compiler/optimizer/InlinerTempForJ9.cpp
Comment (IBMJimmyk) [reply runtime/compiler/optimizer/InlinerTempForJ9.cpp:2087]: I'll add the comment.

File: runtime/compiler/x/codegen/J9TreeEvaluator.cpp
Comment (IBMJimmyk) [reply runtime/compiler/x/codegen/J9TreeEvaluator.cpp:10093]: I can do that. I'll also change it in my OMR side changes as well.

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp
Comment (IBMJimmyk) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp:12223]: I'll do this.

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp
Comment (IBMJimmyk) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp:8176]: This can be done.

**Comment by IBMJimmyk** [comment]
I have some more tests that look like they are failing now. So I'll need to figure out the problem before this can be merged.

**Comment by IBMJimmyk** [comment]
I found the problem.

https://github.com/eclipse/omr/blob/8c233cdf9151d5b097733a426537d4fd6c37051f/compiler/optimizer/VPHandlers.cpp#L5410

`isUnsafeCAS()` does not pass in a parameter.

https://github.com/eclipse/omr/blob/8c233cdf9151d5b097733a426537d4fd6c37051f/compiler/compile/OMRMethod.hpp#L153

So NULL is passed in as the `TR::Compilation *`.

https://github.com/IBMJimmyk/openj9/blob/8f7d8c2ce7914c5ae4a80413002fa8221077f7cb/runtime/compiler/env/j9method.cpp#L5545-L5548

This is the temporary platform check that is here until support is added for every platform. It assumed `TR::Compilation * c` is not NULL which is incorrect.

I will double check these platform checks do not cause any other problems.

**Comment by IBMJimmyk** [comment]
I retested after I added the fix to `isUnsafeCAS` so it doesn't use a NULL comp object and the tests pass now.

@hzongaro Can you look over the latest changes? If they are good I think this is finally ready to be merged in.

**Review by hzongaro — APPROVED** [review]
Looks good.  I just have one suggestion for a potential clean up item for a follow on pull request.

**Comment by hzongaro** [comment]
May I ask you to remove "WIP:" from the titles of this pull request and eclipse/omr#7438?

**Comment by hzongaro** [comment]
Sorry for the late comment.  May I ask you to reformat your commit comment so that lines do not exceed 72 characters in length, per https://github.com/eclipse-openj9/openj9/blob/master/CONTRIBUTING.md#commit-guidelines?

There is also a typo in the commit comment:  "differtiate" -> "differentiate"

**Comment by hzongaro** [comment]
Jenkins test sanity,sanity.openjdk all jdk8,jdk11,jdk17,jdk21

**Comment by hzongaro** [comment]
Sorry - the testing I initiated will fail because I forgot to include the `depends eclipse/omr#7438` clause.

File: runtime/compiler/env/j9method.cpp
Comment (IBMJimmyk) [reply runtime/compiler/env/j9method.cpp:5536]: Yes. I can do that as well.

**Comment by IBMJimmyk** [comment]
I fixed the commit message.

**Comment by hzongaro** [comment]
Jenkins test sanity,sanity.openjdk all jdk8,jdk11,jdk17,jdk21 depends eclipse/omr#7438

**Comment by IBMJimmyk** [comment]
The Windows failure looks to be this:
```
[WS-CLEANUP] Deleting project workspace...
16:39:01  [WS-CLEANUP] Deferred wipeout is disabled by the job configuration...
16:39:19  ERROR: Cannot delete workspace :Unable to delete 'F:\Users\jenkins\workspace\Build_JDK21_x86-64_windows_Personal\openssl\NUL'. Tried 3 times (of a maximum of 3) waiting 0.1 sec between attempts.
```
I don't think that is related to my change.

**Comment by hzongaro** [comment]
Jenkins test sanity,sanity.openjdk windows jdk21 depends eclipse/omr#7438

**Comment by IBMJimmyk** [comment]
I'll take a look at the sanity.functional failures for Windows.

**Comment by hzongaro** [comment]
Jenkins test sanity,sanity.openjdk win jdk21 depends eclipse/omr#7438

**Comment by hzongaro** [comment]
Jenkins test sanity.openjdk aix jdk17 depends eclipse/omr#7438

**Comment by hzongaro** [comment]
Jenkins test sanity.openjdk aix jdk17 depends eclipse/omr#7438

**Comment by IBMJimmyk** [comment]
All of the JDK8 Windows failures fall into the same pattern. The test workload itself is successful and prints the expected result for the test and shortly afterwards there is a segmentation fault in `j9shr29.dll`.

Here is an example from `SCURLClassLoaderTests_SE80_0`
```
[2024-09-24T02:17:44.442Z] Testing: FindStore.FindExplicit
[2024-09-24T02:17:44.442Z] Test start time: 2024/09/23 22:17:43 Eastern Standard Time
[2024-09-24T02:17:44.442Z] Running command: C:/Users/jenkins/workspace/Test_openjdk8_j9_sanity.functional_x86-32_windows_Personal_testList_0/jdkbinary/j2sdk-image\jre\bin\java.exe  -Xjit -Xgcpolicy:gencon -Xnocompressedrefs    -Xshareclasses:persistent,controlDir=.,name=FindStore,noTimestampChecks -cp .\FindStore\Nothing.jar;.\FindStore\A.jar;.\FindStore\B.jar;.\FindStore\C.jar;.\FindStore\D.jar jnurlcldr.shared.findstore.A_Main
[2024-09-24T02:17:44.442Z] Time spent starting: 0 milliseconds
[2024-09-24T02:17:44.442Z] Time spent executing: 328 milliseconds
[2024-09-24T02:17:44.442Z] Test result: FAILED
[2024-09-24T02:17:44.442Z]  [OUT] CACHED A
[2024-09-24T02:17:44.442Z]  [OUT] CACHED B
[2024-09-24T02:17:44.442Z]  [OUT] CACHED C
[2024-09-24T02:17:44.442Z]  [OUT] CACHED D
[2024-09-24T02:17:44.442Z]  [OUT] Result=3
[2024-09-24T02:17:44.442Z]  [ERR] Unhandled exception
[2024-09-24T02:17:44.442Z]  [ERR] Type=Segmentation error vmState=0x00000000
[2024-09-24T02:17:44.442Z]  [ERR] J9Generic_Signal_Number=00000004 ExceptionCode=c0000005 ExceptionAddress=706141DA ContextFlags=0001007f
[2024-09-24T02:17:44.442Z]  [ERR] Handler1=71625EA0 Handler2=715454D0 InaccessibleWriteAddress=22A56118
[2024-09-24T02:17:44.442Z]  [ERR] EDI=013C2FC8 ESI=227E56D8 EAX=22A55FE8 EBX=2298F2B8
[2024-09-24T02:17:44.442Z]  [ERR] ECX=FFFFFFFF EDX=010D2000
[2024-09-24T02:17:44.442Z]  [ERR] EIP=706141DA ESP=01B4F41C EBP=22FE5538 EFLAGS=00210246
[2024-09-24T02:17:44.442Z]  [ERR] GS=002B FS=0053 ES=002B DS=002B
[2024-09-24T02:17:44.442Z]  [ERR] Module=C:\Users\jenkins\workspace\Test_openjdk8_j9_sanity.functional_x86-32_windows_Personal_testList_0\jdkbinary\j2sdk-image\jre\bin\default\j9shr29.dll
[2024-09-24T02:17:44.442Z]  [ERR] Module_base_address=70610000 Offset_in_DLL=000041da
[2024-09-24T02:17:44.442Z]  [ERR] Target=2_90_20240923_809 (Windows Server 2019 10.0 build 17763)
[2024-09-24T02:17:44.442Z]  [ERR] CPU=x86 (4 logical CPUs) (0x3fff77000 RAM)
[2024-09-24T02:17:44.442Z]  [ERR] ----------- Stack Backtrace -----------
[2024-09-24T02:17:44.442Z]  [ERR] J9VMDllMain+0x31da (0x706141DA [j9shr29+0x41da])
[2024-09-24T02:17:44.442Z]  [ERR] J9VMDllMain+0x9aa (0x706119AA [j9shr29+0x19aa])
[2024-09-24T02:17:44.442Z]  [ERR] ---------------------------------------
[2024-09-24T02:17:44.442Z]  [ERR] JVMDUMP039I Processing dump event "gpf", detail "" at 2024/09/23 22:17:43 - please wait.
[2024-09-24T02:17:44.442Z]  [ERR] JVMDUMP013I Processed dump event "gpf", detail "".
[2024-09-24T02:17:44.442Z] >> Success condition was found: [Output match: Result=3]
[2024-09-24T02:17:44.442Z] >> Failure condition was not found: [Output match: LOCAL]
[2024-09-24T02:17:44.442Z] >> Failure condition was found: [Output match: Unhandled Exception]
[2024-09-24T02:17:44.442Z] >> Failure condition was not found: [Output match: Exception:]
[2024-09-24T02:17:44.442Z] >> Failure condition was not found: [Output match: Error:]
```
The test successfully prints out `Result=3` which is the expected result of a successful test. Shortly afterwards, there is a segfault in `j9shr29.dll`.

I have tried to reproduce the problem but have not been successful. When the tests were run as a part of this PR, this error occurred about 12 times total. I tried rerunning the tests 5 times each with a build with and without my changes and they all passed.

I stumbled upon this similar looking problem:
https://github.com/eclipse-openj9/openj9/issues/20227

In that issue, there is also a test with a successful result followed by a crash in `j9shr29.dll`. But, the backtrace isn't quite the same. The problem in that issue was fixed and the change merged in after this build was kicked off. So the testing of my changes does not include the fix.

I think by builds go up to here in OpenJ9:
https://github.com/eclipse-openj9/openj9/commits/16ab797349b9bda5a5de2a9ff27ef2d932a4716a/ (Sept 23)
So it didn't include: https://github.com/eclipse-openj9/openj9/pull/20196/commits (merged sept 24)

**Comment by IBMJimmyk** [comment]
I was able to reproduce the JDK8 Windows failures on a build that excluded both my Unsafe CAS work and the `j9shr29.dll` crash fix. So I can say those failures are unrelated to my change and also already fixed in newer builds.

@hzongaro Are there any other concerns that need to be addressed?

**Comment by hzongaro** [comment]
Failures are due to known issues.  Performing a coordinated merge with eclipse/omr#7438

**Pull request merged**
